clayui.com/src/components/IconSearch/index.js

```diff
@@ -0,0 +1,84 @@
+/**
+ * © 2020 Liferay, Inc. <https://liferay.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import ClayForm, {ClayInput} from '@clayui/form';
+import ClayIcon from '@clayui/icon';
+import React, {useMemo, useState} from 'react';
+
+const spritemap = '/images/icons/icons.svg';
+
+/**
+ * Component that allows searching through icons
+ * @param {string} label - Label of the search input field.
+ * @param {string} placeholder - Placeholder of the search input field.
+ * @param {Array[Object]} source - Path to the source JSON. Needs to follow pattern of [{name: String, aliases: Array<String>}]
+ * @param {string} type - Type of icons being provided, supports "icons" and "flags"
+ */
+const IconSearch = ({
+	label = 'Search Icons',
+	placeholder = 'Search Icons...',
+	source,
+	type = 'icons',
+}) => {
+	const [searchQuery, setSearchQuery] = useState('');
+	let list = [];
+
+	const filteredIcons = useMemo(() => {
+		const query = searchQuery.toLowerCase();
+
+		return source.filter(
+			({aliases, name}) =>
+				name.toLowerCase().includes(query) ||
+				aliases.some(alias => alias.toLowerCase().includes(query))
+		);
+	}, [searchQuery]);
+
+	if (filteredIcons.length) {
```

Comment:
I think you could simplify this if/else block slightly since the `.filter()` should always return an array.

```js
list = filteredIcons.length ? filteredIcons : source
```

Reply:
I don't think we can, since we want to have an empty array in the case where the `searchQuery` doesn't match anything. The logic being if `searchQuery` exists but doesn't match anything return an empty array instead of `source`, which is what your proposal would do. Resulting in no icons matching the filter but all icons being shown.

If `filteredIcons.length` is falsy, that means no icons match the filter, in that case we want to show the `no results` message, not the `source`.

Reply:
Oh I see what you mean, then for that case we could do `list = searchQuery ? filteredIcons : source`

This should account for

`searchQuery` exists and matches no icons  === `[]` or `filteredIcons`(should be empty array if nothing matches)
`searchQuery` exists and matches icons  === `filteredIcons`
`searchQuery` does not exist  === source

unless I am going crazy and just can't think straight...